designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+``sam build`` command
+=====================
+This is the design for a command to **build** a Lambda function. **Build** is the operation of converting the function's
+source code to an artifact that can be executed on AWS Lambda.
+
+
+What is the problem?
+--------------------
+To run a function on AWS Lambda, customers need to provide a zip file containing an executable form of the code. The
+process of creating the executable usually involves downloading dependent libraries, optionally compiling the code
+on Amazon Linux, copying static assets, and arranging the files in a directory structure that AWS Lambda accepts.
+In some programming languages like Javascript, this process is fairly straightforward (just zip a folder), and in
+others like Python, it is much involved.
+
+Customers using SAM CLI can easily create Lambda function code (using ``sam init``) and package their artifact as a
+zip file (using ``sam package``), but they have to handle the build process themselves. Customers usually implement
+their own build scripts or adopt other's scripts from the internet. This is where many customers fall off the cliff.
+
+
+What will be changed?
+---------------------
+In this proposal, we will be providing a new command, ``sam build``, to build Lambda functions for all programming
+languages that AWS Lambda supports. The cardinality of the problem is number of programming languages (N) times the
+number of package managers (M) per language. Hence is it is nearly impossible to natively support each combination.
+Instead, ``sam build`` will support an opinionated set of package managers for all programming languages. We will
+provide an option for customers to bring their own build commands or override the default for any programming language.
+SAM CLI will still take care of the grunt work of iterating through every function in the SAM template, figuring out
+the source code location, creating temporary folders to store built artifacts, running the build command and
+move artifacts to right location.
+
+
+Success criteria for the change
+-------------------------------
+#. Support all programming languages supported by AWS Lambda
+
+   * Nodejs with NPM
+   * Java with Maven
+   * Python with PIP
+   * Golang with Go CLI
+   * Dotnetcore with DotNet CLI
+
+
+#. Each Lambda function in SAM template gets built
+
+#. Produce stable builds (best effort): If the source files did not change, built artifacts should not change.
+
+#. Built artifacts should "just work" with ``sam local`` and ``sam package`` suite of commands
+
+#. Ability to provide arbitrary build commands for each Lambda Function runtime. This will either override the built-in
+   default or add build support for languages/tools that SAM CLI does not support (ex: java+gradle).
+
+#. Opt-in to building native dependencies that can run on AWS Lambda using Docker.
+
+#. Support one dependency manifest (ex: package.json) entire app or one per each Lambda function.
```

Comment:
I'm unaware of a good use case for having a global dependency manifest instead of per-function manifests. Per-function manifests is a best practice:

* Reduces package sizes (which has lots of benefits like reduced cold-start latencies)
* Reduces security attack space (more deps means more potential vulnerabilities in each function)
* Reduces confusion (why is *this* module in *that* function)
* Reduces dependency resolution issues (in case you mix global and per-function manifests)
* and more I haven't thought of off the top of my head...

The only use case for global manifests that I can see is that you only need to update one manifest to update dependencies that all functions use. But the burden of updating each manifest individually is trivial. Simple tooling, if necessary, can remove this burden.

Reply:
The 50MB limit per package isn't much anyway, so by sharing dependencies you're limiting what a single SAM template can do unnecessarily. 

For node.js projects at least, I'd prefer separate package.json per lambda if there are multiple lambdas in a project. in fact, it's actually a separate package.json per project (directory), so a single directory can host multiple lambdas. that way, if someone wants to speed up build time by sharing dependencies, great, they can do it. if they want to keep it minimal per lambda (makes more sense performance-wise), they can.

Reply:
@txase @gojko  Thanks for the detailed comments. This makes a lot of sense. I will remove the app-level manifest support for now. We can always add it later if ppl have a real good use-case for it. May be make it opt-in at that point.